### DISCUSSION TITLE
Does realm actually save memory if properties of realm objects have backing fields in anyway?

### DISCUSSION THREAD

**Discussion by daawaan4x** [realm/realm-dotnet#2738] [Q&A] [answered]
### Realm Properties and Zero-Copy Architecture
I've just realized this recently. Realm has backing fields for properties in case the object is used as a standalone object instead of a managed one. In this way, the Realm object can be reused like a normal POCO. Once the Realm object is managed, these backing fields won't be used by the properties anymore since the properties will now point to its corresponding value on the disk. Consequently, the backing fields will still occupy memory despite being unused. Since realm's objective is to have a zero-copy architecture to minimize memory usage, doesn't this violate that objective?

### Questions for Clarifications
Since I lack knowledge on the capabilities of assembly weavers, my statements above may not be true. How is memory technically conserved in the current version of this SDK? and if there's still further memory that can be conserved, I would also like to ask if there is already a solution that is currently being planned once source generators will replace the assembly weavers.

**Comment by papafe** [comment, marked as answer]
Hi @daawaan, there are different points in your question, so I'll try to answer to all of them:

- The zero-copy architecture of Realm means that there will be zero copies from the database, and it doesn't means that there will be zero allocations of backing fields. So that does not violate our main objective, that is to minimise expensive operations, such as reading data from disk and copying it to memory. 
- Another thing to notice is that primitive values such as integers, booleans and so on don't really take much space in memory, so even if those are unused they don't contribute much to the issue. The problem is with "heavier" types, such as strings, collections and objects, as they can take an arbitrary amount of space. But those are actually never used while the realm object is managed, as properties directly access the memory in that case, and so they don't contribute. 
- The way that memory is conserved is by having the RealmObjects directly access the properties in memory, instead of copying them. If you think about a SQL database, when a value is returned from a query, then the properties are copied from the disk to the memory so that it can be accessed. To make an example, if you execute 50 times a query that returns exactly the same object from the database with SQL, then the memory occupied will be about 50 times the memory of a single object. If you execute that in Realm, then the memory occupied will be marginally higher than the memory occupied for a single object, because all those 50 objects access directly the same locations in memory, and nothing is copied.
- Even when we will have source generators I don't think that our approach will change drastically, as we still want to give developers the possibility of using POCOs as you said, and it's not something that can be done without using backing fields.

I hope this answer your questions! If something is not clear feel free to ask 😄

**Reply by nirinchev** [reply]
And to expand on the last point - we're definitely looking into source generators and hope to experiment some more with the API early next year.
